accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+8. npm CLI verifies that the certificate used with the Rekor npm package entry is signed with a certificate matching npm and the package name.
+9. npm CLI verifies that the certificate used to sign the package digest matches the package name and repository in the build provenance’s config source section.
+
+There are two obvious failure scenarios we need to handle when verifying:
+- The network call to fetch attestations fails: This should probably error and instruct users to re-run `npm audit signatures` or `npm install`.
+- Something genuinely unexpected happened and the attestation signature doesn't match the package that's been downloaded: This should fail the audit or install instruct users to manually inspect the failing package(s) to determine if they might have been tampered with.
+
+### Supporting third-party npm clients
+Third-party npm clients, e.g. `yarn` and `pnpm`, should be able to implement build provenance generation using the [sigstore-js](https://github.com/sigstore/sigstore-js) client library. This library will contain all the functionality required to extract build provenance information, generate the OIDC id token on supported CI/CD systems, authenticate it against Fulcio, generate the signature and upload it to the public Rekor ledger.
+
+We're planning on providing detailed documentation on how to wire up the Sigstore client library in any npm client.
+
+### Supporting third-party npm registries
+At a high level, third-party registries will need to support the following:
+- Generate release attestations
+- Accept build provenance attestation documents on publish, alongside the packument (`package.json`) and tarball
+- Storing the attestations in the registry
+- Providing a query endpoint for retrieving attestations
+
+We're planning on providing detailed documentation on the required changes.
+
+### Non-falsifiable provenance using a trusted builder
+To ensure the provenance information hasn't been tampered with during the build, e.g. by a malicious dependency, we need to have a way to run a build and publish in a trusted build environment that can't be directly accessed or modified by the source repository where it's being executed.
+
+Non-falsifiable provenance information is a [requirement to reach SLSA level 3](https://slsa.dev/spec/v0.1/requirements#non-falsifiable) (framework for securing the supply chain).
+
+Using a trusted builder to publish an npm release will ensure a higher level of build integrity. Using it will be optional, but highly recommended. Verification of build provenance will include checks to determine if a trusted builder was used.
+
+At a high level, the trusted builder will checkout the git repo, install dependencies and execute `npm run build` followed by `npm publish`. Maintainers can only control what is run during `npm run build` not add or remove steps from the trusted builder. Provenance generation will be run in a separate workflow step that's isolated from the build and install.
```

Comment:
`npm run build` here should be configurable. This is not the build command that everyone uses, and could potentially add significant overhead for projects with existing workflows.

Reply:
Only run `npm publish` and allow extension via `prepublishOnly`.

Reply:
… or prepare, or prepack, or any other relevant lifecycle event, presumably?